File: 1367.linked-list-in-binary-tree.py
#
# @lc app=leetcode id=1367 lang=python3
#
# [1367] Linked List in Binary Tree
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubPath(self, head: Optional[ListNode], root: Optional[TreeNode]) -> bool:
        def check(node: Optional[TreeNode], ln: Optional[ListNode]) -> bool:
            if not ln:
                return True
            if not node or node.val != ln.val:
                return False
            return check(node.left, ln.next) or check(node.right, ln.next)
        
        def dfs(node: Optional[TreeNode]) -> bool:
            if not node:
                return False
            if check(node, head):
                return True
            return dfs(node.left) or dfs(node.right)
        
        return dfs(root)
# @lc code=end

File: 1367.linked-list-in-binary-tree.cpp
#
# @lc app=leetcode id=1367 lang=cpp
#
# [1367] Linked List in Binary Tree
#

# @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSubPath(ListNode* head, TreeNode* root) {
        if (!root) return false;
        // Check if path starts at current root
        if (dfs(head, root)) return true;
        // Otherwise try left and right subtrees
        return isSubPath(head, root->left) || isSubPath(head, root->right);
    }
    
    // Helper to check if linked list matches starting at this tree node
    bool dfs(ListNode* head, TreeNode* root) {
        if (!head) return true; // All nodes matched
        if (!root) return false; // Tree ends before list ends
        if (head->val != root->val) return false;
        // Current values match, continue with next node in list and children
        return dfs(head->next, root->left) || dfs(head->next, root->right);
    }
};
# @lc code=end